File: python/leetcodehard/largest_rectangle_histogram.py
"""
https://leetcode.com/problems/largest-rectangle-in-histogram/

status - completed
"""

from base_logger.logging_event import create_logger


class Solution:

    log = create_logger(__name__)

    def largestRectangleArea(self, heights):
        stack = []
        max_area = 0

        for i, height in enumerate(heights + [0]):
            while stack and heights[stack[-1]] >= height:
                top = stack.pop()
                width = i if not stack else i - stack[-1] - 1
                max_area = max(max_area, heights[top] * width)
            stack.append(i)

        return max_area


if __name__ == "__main__":
    test_input_1 = [2, 1, 5, 6, 2, 3]
    output = Solution().largestRectangleArea(test_input_1)
    Solution().log.info("Output is %s", output)

    test_input_2 = [2, 4]
    output = Solution().largestRectangleArea(test_input_2)
    Solution().log.info("Output is %s", output)

File: typescript/leetcodehard/largestRectangleHistogram.ts
/*
https://leetcode.com/problems/largest-rectangle-in-histogram/

status - completed
*/

import { logp } from "../utils/logger";

function largestRectangleArea(heights: number[]): number {
  const stack: number[] = [];
  let maxArea = 0;

  for (let i = 0; i <= heights.length; i++) {
    const h = i === heights.length ? 0 : heights[i];

    while (stack.length > 0 && heights[stack[stack.length - 1]] >= h) {
      const height = heights[stack.pop()!];
      const width = stack.length === 0 ? i : i - stack[stack.length - 1] - 1;
      maxArea = Math.max(maxArea, height * width);
    }

    stack.push(i);
  }

  return maxArea;
}

logp(`input [2,1,5,6,2,3] expected 10 actual ${largestRectangleArea([2, 1, 5, 6, 2, 3])}`);
logp(`input [2,4] expected 4 actual ${largestRectangleArea([2, 4])}`);
